File: shell/gnome/src/popup/statusNotice.js
import Clutter from 'gi://Clutter';
import St from 'gi://St';
import { _ } from '../i18n.js';
import { providerIncident } from '../providerStatus.js';
import { button, column, label, row, spacer, themeIcon, wrappingLabel } from '../widgets.js';
import { popupIcon } from './icons.js';
import {
    statusAge,
    statusDetail,
    statusIcon,
    statusKind,
    statusStarted,
    statusTitle,
    statusTone,
} from './statusTexts.js';

function statusLink(ctx, urlOf) {
    const content = row({ style_class: 'headroom-status-link-box' });
    content.add_child(label(_('Status page'), 'headroom-status-link-text'));
    content.add_child(popupIcon(ctx.dir, 'external', 'headroom-status-link-icon'));
    const actor = button(content, 'headroom-status-link', () => {
        const url = urlOf();
        if (url) ctx.actions.openUrl(url);
    });
    return actor;
}

function iconTile(status) {
    const tile = new St.Bin({
        style_class: `headroom-notice-tile ${statusTone(status)}`,
        y_align: Clutter.ActorAlign.START,
    });
    tile.set_child(themeIcon(statusIcon(status), 'headroom-notice-icon'));
    return tile;
}

class FullNotice {
    constructor(ctx, status, urlOf) {
        this.actor = row({ style_class: `headroom-notice ${statusTone(status)}`, x_expand: true });
        const texts = column({ style_class: 'headroom-notice-texts', x_expand: true });
        this._title = wrappingLabel('', 'headroom-notice-title');
        this._detail = wrappingLabel('', 'headroom-notice-detail');
        const meta = row({ style_class: 'headroom-notice-meta' });
        this._started = label('', 'headroom-notice-detail');
        meta.add_child(this._started);
        meta.add_child(spacer());
        meta.add_child(statusLink(ctx, urlOf));
        for (const actor of [this._title, this._detail, meta]) texts.add_child(actor);
        this.actor.add_child(iconTile(status));
        this.actor.add_child(texts);
    }

    update(status, now) {
        this._title.text = statusTitle(status);
        const detail = statusDetail(status);
        this._detail.text = detail ?? '';
        this._detail.visible = detail !== null;
        this._started.text = statusStarted(status, now) ?? '';
    }
}

class CompactNotice {
    constructor(ctx, status, urlOf) {
        this.actor = row({ style_class: 'headroom-status-line', x_expand: true });
        this.actor.add_child(
            new St.Widget({ style_class: `headroom-status-dot ${status.tone}`, y_align: Clutter.ActorAlign.CENTER })
        );
        this._kind = label('', 'headroom-status-kind');
        this._age = label('', 'headroom-status-age');
        for (const actor of [this._kind, this._age, spacer(), statusLink(ctx, urlOf)]) this.actor.add_child(actor);
        ctx.tooltips.attach(this._kind, () => (this._status ? statusTitle(this._status) : null));
    }

    update(status, now) {
        this._status = status;
        this._kind.text = statusKind(status);
        this._age.text = statusAge(status, now) ?? '';
    }
}

export class StatusNotice {
    constructor(ctx, provider) {
        this._ctx = ctx;
        this._provider = provider;
        const status = this._status();
        const urlOf = () => this._status()?.url ?? null;
        const Kind = ctx.display.density === 'compact' ? CompactNotice : FullNotice;
        this._view = new Kind(ctx, status, urlOf);
        this.actor = this._view.actor;
        this.update();
    }

    _status() {
        return providerIncident(this._ctx.providerStatus(), this._provider);
    }

    update() {
        const status = this._status();
        if (status) this._view.update(status, this._ctx.now());
    }
}

File: crates/headroom-tray/src/ui/status_notice.rs
use gtk::prelude::*;
use jiff::Timestamp;

use crate::popup_model::status::{StatusView, elapsed_text, started_text};
use crate::ui::breakdown::dot;
use crate::ui::context::{Action, Ctx};
use crate::ui::motion::{SLIDE_MS, fade};
use crate::ui::widgets::{button, column, icon, label, row, spacer, wrapping_label};

const TILE_ICON: i32 = 14;
const LINK_ICON: i32 = 10;

fn link(ctx: &Ctx, url: &str, class: &str) -> gtk::Button {
    let content = row(3, &[]);
    content.append(&label(
        ctx.locale.lang.tr("Status page"),
        &["headroom-status-link-text", class],
    ));
    let arrow = icon(
        "adw-external-link-symbolic",
        LINK_ICON,
        &["headroom-status-link-icon"],
    );
    arrow.set_valign(gtk::Align::Center);
    content.append(&arrow);
    let link = button(
        &content,
        &["headroom-status-link"],
        ctx.action(Action::OpenUrl(url.to_owned())),
    );
    link.set_tooltip_text(Some(url));
    link.set_valign(gtk::Align::Center);
    link
}

fn ticking(
    ctx: &Ctx,
    text: &gtk::Label,
    since: Timestamp,
    show: fn(crate::i18n::Lang, Timestamp, Timestamp) -> String,
) {
    let (lang, text) = (ctx.locale.lang, text.clone());
    ctx.on_tick(move |now| text.set_text(&show(lang, since, now)));
}

fn tile(view: &StatusView) -> gtk::Box {
    let tile = row(0, &["headroom-notice-tile"]);
    tile.set_valign(gtk::Align::Start);
    let name = if view.critical {
        "dialog-error-symbolic"
    } else {
        "dialog-warning-symbolic"
    };
    let image = icon(name, TILE_ICON, &[]);
    image.set_hexpand(true);
    image.set_halign(gtk::Align::Center);
    tile.append(&image);
    tile
}

fn full_notice(ctx: &Ctx, view: &StatusView, now: Timestamp) -> gtk::Box {
    let kind = if view.critical { "error" } else { "warning" };
    let body = row(8, &["headroom-notice", "headroom-status-notice", kind]);
    body.append(&tile(view));
    let texts = column(1, &[]);
    texts.set_hexpand(true);
    texts.append(&wrapping_label(&view.title, &["headroom-notice-title"]));
    if let Some(detail) = &view.detail {
        texts.append(&wrapping_label(detail, &["headroom-notice-detail"]));
    }
    let meta = row(5, &["headroom-status-meta"]);
    if let Some(since) = view.started_at {
        let started = label(
            &started_text(ctx.locale.lang, since, now),
            &["headroom-notice-detail"],
        );
        started.set_ellipsize(gtk::pango::EllipsizeMode::End);
        ticking(ctx, &started, since, started_text);
        meta.append(&started);
    }
    meta.append(&spacer());
    meta.append(&link(ctx, &view.url, "full"));
    texts.append(&meta);
    body.append(&texts);
    body
}

fn compact_notice(ctx: &Ctx, view: &StatusView, now: Timestamp) -> gtk::Box {
    let line = row(7, &["headroom-status-line"]);
    let tone = if view.critical { "crit" } else { "notice" };
    line.append(&dot(ctx.color(tone)));
    line.append(&label(&view.kind, &["headroom-status-line-title"]));
    if let Some(since) = view.started_at {
        let elapsed = label(
            &format!("· {}", elapsed_text(ctx.locale.lang, since, now)),
            &["headroom-status-line-meta"],
        );
        let (lang, shown) = (ctx.locale.lang, elapsed.clone());
        ctx.on_tick(move |now| shown.set_text(&format!("· {}", elapsed_text(lang, since, now))));
        line.append(&elapsed);
    }
    line.append(&spacer());
    line.append(&link(ctx, &view.url, "compact"));
    line.set_tooltip_text(Some(&view.title));
    line
}

pub fn status_notice(ctx: &Ctx, view: &StatusView, now: Timestamp, animate: bool) -> gtk::Box {
    let notice = if ctx.compact() {
        compact_notice(ctx, view, now)
    } else {
        full_notice(ctx, view, now)
    };
    if animate {
        fade(&notice, 0.0, 1.0, SLIDE_MS, ctx.motion);
    }
    notice
}

pub fn status_mark(view: &StatusView) -> gtk::Image {
    let (name, class) = if view.critical {
        ("dialog-error-symbolic", "headroom-header-critical")
    } else {
        ("dialog-warning-symbolic", "headroom-header-warning")
    };
    let mark = icon(name, 12, &[class]);
    mark.set_tooltip_text(Some(&view.title));
    mark.set_valign(gtk::Align::Center);
    mark
}
